src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -4280,18 +4193,24 @@ private String layoutChars(boolean sci) {
             // currency fast path
             int lowInt = (int)intCompact % 100;
             int highInt = (int)intCompact / 100;
-            return (Integer.toString(highInt) + '.' +
-                    StringBuilderHelper.DIGIT_TENS[lowInt] +
-                    StringBuilderHelper.DIGIT_ONES[lowInt]) ;
+            int highIntSize = DecimalDigits.stringSize(highInt);
+            byte[] buf = new byte[highIntSize + 3];
+            DecimalDigits.putPairLatin1(buf, highIntSize + 1, lowInt);
+            buf[highIntSize] = '.';
+            DecimalDigits.getCharsLatin1(highInt, highIntSize, buf);
+            try {
+                return JLA.newStringNoRepl(buf, StandardCharsets.ISO_8859_1);
+            } catch (CharacterCodingException cce) {
+                throw new AssertionError(cce);
+            }
         }
 
-        StringBuilderHelper sbHelper = new StringBuilderHelper();
         char[] coeff;
         int offset;  // offset is the starting index for coeff array
         // Get the significand as an absolute value
         if (intCompact != INFLATED) {
-            offset = sbHelper.putIntCompact(Math.abs(intCompact));
-            coeff  = sbHelper.getCompactCharArray();
+            coeff = new char[19];
+            offset = DecimalDigits.getChars(Math.abs(intCompact), coeff.length, coeff);
```

Comment:
`StringBuilderHelper.putIntCompact` asserts that `Math.abs(intCompact)` is non-negative, but there is no such assert in `DecimalDigits.getChars`. I'm not too familiar with `BigDecimal` and maybe `intCompact` wouldn't be set for `Long.MIN_VALUE`. If so perhaps this should use `StrictMath#absExact(long)` for clarity rather than lean on an assert?

Reply:
```java
static final long INFLATED = Long.MIN_VALUE;
```

INFLATED is Long.MIN_VALUE, and a pre-judgment is made here
```java
if (intCompact != INFLATED) {
      offset = DecimalDigits.getChars(Math.abs(intCompact), coeff.length, coeff);
}
```
So Math.abs(intCompact) here will not have overflow problems